### ISSUE TITLE
[HOLD for payment 2023-11-21][$1000] FAB - After a refresh, clicking back button on the 'I am a teacher' page does not take you back to the previous page

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#26282]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Click on FAB
2. Click **Save the World** > **I am a teacher**
3. Click on the **Update email address** button
4. Refresh page.
5. Click on the back button

## Expected Result:
Clicking back takes you to the 'I am a teacher' page

## Actual Result:
Clicking back take you to profile page

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** v1.3.59-0

**Reproducible in staging?:** Y

**Reproducible in production?:** N

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/115492554/6fa35df5-66ca-4bf5-8a73-4dbf63cf5030

https://github.com/Expensify/App/assets/115492554/97adfcb6-6f0f-490b-86e5-155f37c74374

**Expensify/Expensify Issue URL:**

**Issue reported by:** @daveSeife

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1693399672874879

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01e00165b404e4216b</li>
        <li>Upwork Job ID: 1696904936518942720</li>
        <li>Last Price Increase: 2023-10-30</li>
<li>Automatic offers: </li>
<ul>
<li>jjcoffee | Reviewer | 27440202</li>
<li>paultsimura | Contributor | 27440206</li>
<li>daveSeife | Reporter | 27440209</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @trjExpensify</details>

**Comment by paultsimura** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
After refreshing a page on "I am a teacher > Update email" drawer, the "Go Back" button returns to Profile settings

### What is the root cause of that problem?
On refresh, the navigation has no information about the previous page, so the goBack function navigates to the fallback route: `ROUTES.SETTINGS_PROFILE`

https://github.com/Expensify/App/blob/d8b9cddf4c91d2d599707c15b771d7b8249ca4a1/src/pages/settings/Profile/Contacts/ContactMethodsPage.js#L117

### What changes do you think we should make in order to solve the problem?
Pass a URL parameter `backTo` while opening the "Edit Email" page, and use it as a fallback route for the "Go back" button:
```
const navigateBackTo = lodashGet(props.route, 'params.backTo', '');
...
<HeaderWithBackButton
                ...
                onBackButtonPress={() => Navigation.goBack(navigateBackTo)}
            />
```


### What alternative solutions did you explore? (Optional)

**Comment by conorpendergrast** [comment]
FYI @youssef-lr @StefStavri 

I don't think this needs to be an Hourly or Deploy Blocker because this is a beta feature that's part of wave7. I'll change those labels now, but change them back if you disagree!

**Comment by conorpendergrast** [comment]
I've also cross-posted to [our internal channel for this project](https://expensify.slack.com/archives/C02NZ2HGJAZ/p1693405181550689) to make sure that's correct

**Comment by youssef-lr** [comment]
@dhairyasenjaliya could you look into this please?

**Comment by youssef-lr** [comment]
I'm unable to reproduce this..

**Comment by youssef-lr** [comment]
Demoting to NAV and daily as this is a minor bug.

**Comment by conorpendergrast** [comment]
Discussed [internally](https://expensify.slack.com/archives/C02NZ2HGJAZ/p1693408484826479?thread_ts=1693405181.550689&cid=C02NZ2HGJAZ), and as this is so low-impact we'll treat this as NAB and Daily

**Comment by paultsimura** [comment]
Hi, sorry if it's stated somewhere - I couldn't find this policy in the docs and would love to understand:
At the moment when an issue is marked as "Help needed", and an Upwork job is created, shouldn't the proposals be taken into consideration in the "first come - first served" manner?

Or an external contributor still can be assigned manually?

**Comment by hihihass** [comment]
Hy i fixed this bugs :) tomorrow i send explanation and video

**Comment by hihihass** [comment]
Contributor details
Your Expensify account email: jarosius@gmail.com Upwork Profile Link: https://www.upwork.com/freelancers/~01619b3cb9dfb9a540?viewMode=1

**Comment by hihihass** [comment]
i was able to fix this bug: 

https://github.com/Expensify/App/assets/22372335/355d3264-6c3c-4201-be2b-8c92df8e2367

**Comment by hihihass** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
After refreshing a page on "I am a teacher > Update email" drawer, the "Go Back" button returns to Profile settings
### What is the root cause of that problem?
Navigation state is different after page refresh.
### What changes do you think we should make in order to solve the problem?
Save navigation state history in storage.

**Comment by hihihass** [comment]
@izarutskaya  here related bug report by Sofonias Nega
  [6:54 AM](https://expensify.slack.com/archives/C049HHMV9SM/p1693454079049609)

**Comment by conorpendergrast** [comment]
@paultsimura That would typically happen. In this case, @dhairyasenjaliya had been working on [the overall flow](https://github.com/Expensify/App/issues/22857). As a result, it typically makes sense to give them first swing at fixing any bugs that are part of that change

**Comment by paultsimura** [comment]
@conorpendergrast Makes sense, thanks for the explanation.

**Comment by jjcoffee** [comment]
Waiting on @dhairyasenjaliya before reviewing any proposals.

**Comment by hihihass** [comment]
> Waiting on @dhairyasenjaliya before reviewing any proposals.

But bug represented in project wide.

**Comment by hihihass** [comment]
> Hey sorry for delayed in reply actually I will be OOO for while so if anyone up for the fix would be fine but I definitely remember we have fixed or maybe patched similar to this issue so we should look for general solution not only this page since its issue definitely exists with other pages I don't have exact link for the particular fix as currently don't have my mac with me

I can't find any fix in this repo :)

**Comment by conorpendergrast** [comment]
Ok, as @dhairyasenjaliya isn't available, let's keep this open for any proposals (like @hihihass' [above](https://github.com/Expensify/App/issues/26282#issuecomment-1700330108)), FYI @jjcoffee 

@dangrous any disagreement on that? Thanks!

**Comment by jjcoffee** [comment]
It sounds like a general fix is being worked up here #22041 - [asking](https://github.com/Expensify/App/issues/22041#issuecomment-1706642079) there to double check! cc @conorpendergrast

**Comment by dangrous** [comment]
Cool, sounds like we should put this on hold for #22041 then?

**Comment by jjcoffee** [comment]
@dangrous Yup let's do that!

**Comment by conorpendergrast** [comment]
@trjExpensify I'm on parental leave; re-assigning! This has been triaged is held on https://github.com/Expensify/App/issues/22041 as a more general fix

**Comment by dangrous** [comment]
Still holding on https://github.com/Expensify/App/issues/22041

**Comment by dangrous** [comment]
Getting there, still on hold

**Comment by dangrous** [comment]
#22041 was on hold for another issue which is now merged, so we're back to movement. Hopefully soon!

**Comment by trjExpensify** [comment]
@dangrous #22041 is awaiting payment, so both dependencies are now cleared. I've updated the title to remove the hold.

@daveSeife you can retest this on the I am a teacher page and then we should be good to pay out the bug report.

**Comment by daveSeife** [comment]
@trjExpensify I just retested it. The issue is still reproducible.

**Comment by trjExpensify** [comment]
Interesting. Why would that be? CC: @situchan

**Comment by situchan** [comment]
cc: @hoangzinh

**Comment by dangrous** [comment]
Got it, so it sounds like this should stay open to find some new solutions, is that correct? @hoangzinh or @situchan feel free to jump in!

**Comment by hihihass** [comment]
I

> Got it, so it sounds like this should stay open to find some new solutions, is that correct? @hoangzinh or @situchan feel free to jump in!

https://github.com/Expensify/App/issues/26282#issuecomment-1700330108

**Comment by situchan** [comment]
@hihihass no, we don't need to save state in local storage

**Comment by dangrous** [comment]
Given that we have a relatively simple solution, I feel like we might as well fix this - as we did the [issue](https://github.com/Expensify/App/issues/22802) that sparked that other discussion. @jjcoffee do you want to take another look through the proposals? I think @paultsimura's idea seems to sync up with how we fixed that other one. Thanks!

**Comment by jjcoffee** [comment]
In that case, I agree that we can go with @paultsimura's [proposal](https://github.com/Expensify/App/issues/26282#issuecomment-1699252672). We don't want to go down the route of using localStorage as in @hihihass's [proposal](https://github.com/Expensify/App/issues/26282#issuecomment-1700330108).

:ribbon::eyes::ribbon: C+ reviewed

**Comment by dangrous** [comment]
Great, let's do it.

**Comment by paultsimura** [comment]
Thank you all, the [PR](https://github.com/Expensify/App/pull/30608) is ready for review.

**Comment by paultsimura** [comment]
@dangrous the PR is in production already (there was some weird flow: deployed to staging - prod - staging), but the automation didn't update the [HOLD for payment] title, could you please check? 🙏

**Comment by dangrous** [comment]
So that production part is messed up, it's currently only on staging. We should get the right automated message when it's REALLY on production, hopefully!

**Comment by dangrous** [comment]
hrm, you're right - I'll update the title but I'll see if I can get the rest of them to run again (e.g. checklist, etc.)

**Comment by dangrous** [comment]
Okay so no way to get it automatically, annoyingly. @trjExpensify just be aware that we'll have to figure out who needs what at the end of this. I believe it will be @paultsimura as contributor, @jjcoffee as C+, @daveSeife as reporter

And, copying the checklist here:

**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@jjcoffee] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@jjcoffee] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@jjcoffee] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@jjcoffee] Determine if we should create a regression test for this bug.
- [ ] [@jjcoffee] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@trjExpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by jjcoffee** [comment]
- The PR that introduced the bug has been identified. Link to the PR: **N/A - a fallback route was never provided**
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: **N/A**
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: **N/A**
- Determine if we should create a regression test for this bug. **Yes - easy enough to test**
- If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

**Regression Test Proposal**

1. Open FAB and go to Save the World -> I am a teacher
2. Click on the Update email address button
3. Refresh page
4. Click on the back button
5. Verify that you are taken back to the previous screen (i.e. the 'I am a teacher' page)

Do we agree 👍 or 👎

**Comment by trjExpensify** [comment]
Okay, so this is fun!

- Grandfathered base rates on this as the issue was created before the cut off on August 30th with the [first change](https://expensify.slack.com/archives/C01GTK53T8Q/p1693424078129649).
- No #urgency bonus because the PR was created on Oct 30th, after the [second program change](https://expensify.slack.com/archives/C01GTK53T8Q/p1698191269281379) to remove that on Oct 24th. 

By my calculations then: 

- $250 to @daveSeife for the bug report
- $1000 to @jjcoffee for the C+ review
- $1000 to @paultsimura for the bug fix 

As for the regression test. It seems a bit niche in isolation, IMO. Though, @JmillsExpensify @marcochavezf as part of the STW doc wrap-up, have tests been given to Applause for the core flow? Marco, I know you were looking at pointing to a test policy to allow for testing at some point?

**Comment by trjExpensify** [comment]
@jjcoffee @dangrous let me know if you agree on the payouts and I'll action on settling up.

**Comment by dangrous** [comment]
I think that's right!

**Comment by jjcoffee** [comment]
@trjExpensify Yup, I'm good with that!

**Comment by trjExpensify** [comment]
Perfect. Settled up with all three of you. Closing, thanks!

**Comment by marcochavezf** [comment]
> I know you were looking at pointing to a test policy to allow for testing at some point?

Yeah, this is the [GH](https://github.com/Expensify/Expensify/issues/315545) for the QA tests, but we're still [figuring out](https://github.com/Expensify/Web-Expensify/pull/39601) how to flip the policyID and public room reportID for testing.

**Comment by trjExpensify** [comment]
Cool, thanks!
